### PR TITLE
SONARJS-842 Rule: Functions should always return the same type

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<p>Unlike strongly typed languages, JavaScript does not enforce a return type on a function. This means that different paths through a function can
+return different types of values, which can be very confusing to the user and significantly harder to maintain.</p>
+<h2>Noncompliant Code Example</h2>
+<pre>
+function foo(a) {
+  if (a === 1) {
+    return true; // Noncompliant
```
Comment (pynicolas) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.html]: @vilchik-elena Should we move the "//Noncompliant" to the function keyword?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/FunctionReturnTypeCheck.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+        String typesForSecondaryMessage = String.join(" or ", typeNames);
+        issue.secondary(returnStatement.returnKeyword(), "Returns " + typesForSecondaryMessage);
+      }
+    }
+  }
+
+  private static Set<Constraint> normalizeConstraints(Collection<Constraint> constraints) {
+    Set<Constraint> normalizedConstraintSet = new HashSet<>(constraints);
+    normalizedConstraintSet.remove(UNKNOWN_TYPE);
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/FunctionReturnTypeCheck.java]: @vilchik-elena Can't we simply not store `UNKNOWN_TYPE` in `returnedValueConstraintsByReturnStatement`?

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+  "title": "Functions should always return the same type",
+  "type": "CODE_SMELL",
+  "status": "ready",
+  "remediation": {
+    "func": "Linear",
```
Comment (pynicolas) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.json]: @vilchik-elena Shouldn't we add a "cost" when we create an issue?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/FunctionReturnTypeCheck.java (hunk 2 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/FunctionReturnTypeCheck.java]: @pynicolas good idea:)

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.html (hunk 1 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.html]: @pynicolas updated

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.json (hunk 3 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.json]: @pynicolas updated

**Pull request merged**